### PR TITLE
HHH-15725 Criteria API Expression.as adds cast even when the cast type is equal to the expression type

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-15725

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AbstractSqmExpression.java [hunk 1]
```diff
@@ -94,7 +94,15 @@ public SqmExpression<String> asString() {
...
-		return nodeBuilder().cast( this, type );
+		final SqmExpressible<?> expressible = getExpressible();
+		if ( expressible != null && expressible.getExpressibleJavaType().getJavaTypeClass() == type ) {
+			return (SqmExpression<X>) this;
+		}
+		final BasicType<X> basicTypeForJavaType = nodeBuilder().getTypeConfiguration().getBasicTypeForJavaType( type );
+		if ( basicTypeForJavaType == null ) {
+			throw new IllegalArgumentException( "Can't cast expression to unknown type: " + type.getCanonicalName() );
+		}
+		return new AsWrapperSqmExpression<>( basicTypeForJavaType, this );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AbstractSqmExpression.java:105]: Just like we discussed [here](https://github.com/hibernate/hibernate-orm/pull/8653#issuecomment-2222622818), this should always return a new expression instance.

```suggestion
		final BasicType<X> basicTypeForJavaType = nodeBuilder().getTypeConfiguration().getBasicTypeForJavaType( type );
		if ( basicTypeForJavaType == null ) {
			throw new IllegalArgumentException( "Can't cast expression to unknown type: " + type.getCanonicalName() );
		}
		return new AsWrapperSqmExpression<>( basicTypeForJavaType, this );
```

File: hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AsWrapperSqmExpression.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	@Override
+	public void appendHqlString(StringBuilder sb) {
+		expression.appendHqlString( sb );
+	}
+
+	@Override
+	public SqmExpression<T> copy(SqmCopyContext context) {
+		return new AsWrapperSqmExpression<>( getExpressible(), expression.copy( context ) );
+	}
+
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AsWrapperSqmExpression.java:36]:
```suggestion

	@Override
	public <X> SqmExpression<X> as(Class<X> type) {
		return expression.as( type );
	}
```

**Review by beikov — APPROVED** [review]
Might be nice to add code to avoid unnecessary casts if possible, but that's out of scope and something for a different Jira issue :)

**Review by sebersole — APPROVED** [review]
Looks reasonable to me

**Comment by sebersole** [comment]
Leading to all jobs failing

**Comment by dreab8** [comment]
@sebersole looking at the failures

**Pull request merged**